viewing_party/party.py

```diff
@@ -1,23 +1,160 @@
+
+    for movie in user_data_copy["watchlist"]:
+        if movie["title"] == title:
+            user_data_copy["watchlist"].remove(movie)
+            user_data_copy["watched"].append(movie)
+            
+    return user_data_copy
+# ------------- WAVE 2 --------------------
+#Calculates the average rating of all watched movies.
+def get_watched_avg_rating(user_data):
+    if "watched" not in user_data or len(user_data["watched"]) == 0:
+        return 0.0
 
+    total_rating = 0
+    for movie in user_data["watched"]:
+        if "rating" in movie:
+            total_rating += movie["rating"]
+        else:
+            total_rating += 0
+
+    avg_rating = total_rating / len(user_data["watched"])
+    return avg_rating
 
 # -----------------------------------------
+#Determines the most frequently watched genre. 
+def get_most_watched_genre(user_data):
+    if "watched" not in user_data or len(user_data["watched"]) == 0:
+        return None
+
+    genre_count = {}
+    most_watched_genre = None
+    highest_count = 0
+
+    for movie in user_data["watched"]:
+        if "genre" in movie:
+            genre = movie["genre"]
+            if genre in genre_count:
+                genre_count[genre] += 1
+            else:
+                genre_count[genre] = 1
+
+            if genre_count[genre] > highest_count:
+                highest_count = genre_count[genre]
+                most_watched_genre = genre
+
+    return most_watched_genre
+
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
+#Finds movies the user has watched but their friends haven't.
+def get_unique_watched(user_data):
 
-        
+    user_watched = user_data["watched"]
+    
+    friends_watched = []
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.append(movie["title"])
+    
+    unique_movies = []
+    for movie in user_watched:
+        if movie["title"] not in friends_watched:
+                unique_movies.append(movie)
+    
+    return unique_movies
 # -----------------------------------------
+#Finds movies that at least one friend has watched but the user hasn't.
+def get_friends_unique_watched(user_data):
+    user_watched_titles = []
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+
+    friends_unique_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles:
+                if movie not in friends_unique_movies:
+                    friends_unique_movies.append(movie)
+
+    return friends_unique_movies
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
+def get_available_recs(user_data):    
+    #Get the user's streaming subscriptions
+    subscriptions = user_data["subscriptions"]
+
+    friends_unique_movies = get_friends_unique_watched(user_data)

```

Comment:
👍 Invoking a function that you already wrote is a great way to keep this function DRY and concise.